cub/test/catch2_test_device_merge_sort.cu

```diff
@@ -0,0 +1,309 @@
+    custom_val.val = val;
+    return custom_val;
+  }
+};
+
+/**
+ * Generates a shuffled array of key ranks. E.g., for a vector of size 5: [4, 2, 3, 1, 0]
+ */
+template <typename OffsetT>
+thrust::device_vector<OffsetT> make_shuffled_key_ranks_vector(OffsetT num_items, c2h::seed_t seed)
+{
+  thrust::device_vector<OffsetT> key_ranks(num_items);
+  thrust::sequence(key_ranks.begin(), key_ranks.end());
+  thrust::shuffle(
+    key_ranks.begin(), key_ranks.end(), thrust::default_random_engine{static_cast<unsigned int>(seed.get())});
+  return key_ranks;
+}
+
+CUB_TEST("DeviceMergeSort::SortKeysCopy works", "[merge][sort][device]", wide_key_types)
+{
+  using key_t    = typename c2h::get<0, TestType>;
+  using offset_t = std::int32_t;
+
+  // Prepare input
+  // const offset_t num_items = GENERATE_COPY(take(2, random(1, 1000000)), values({500, 1000000, 2000000}));
+  const offset_t num_items = GENERATE_COPY(values({500}));
+  auto key_ranks           = make_shuffled_key_ranks_vector(num_items, CUB_SEED(2));
+  thrust::device_vector<key_t> keys_in(num_items);
+  thrust::transform(key_ranks.begin(), key_ranks.end(), keys_in.begin(), rank_to_key_op_t<offset_t, key_t>{});
+
+  // Perform sort
+  thrust::device_vector<key_t> keys_out(num_items);
+  thrust::fill(keys_out.begin(), keys_out.end(), static_cast<key_t>(42));
+  sort_keys_copy(
+    thrust::raw_pointer_cast(keys_in.data()), thrust::raw_pointer_cast(keys_out.data()), num_items, custom_less_op_t{});
+
+  // Verify results
+  thrust::host_vector<key_t> keys_expected(num_items);
+  auto key_ranks_it = thrust::make_counting_iterator(offset_t{});
+  thrust::transform(key_ranks_it, key_ranks_it + num_items, keys_expected.begin(), rank_to_key_op_t<offset_t, key_t>{});
+
+  REQUIRE(keys_expected == keys_out);
+}
+
+CUB_TEST("DeviceMergeSort::SortKeys works", "[merge][sort][device]", wide_key_types)
+{
+  using key_t    = typename c2h::get<0, TestType>;
+  using offset_t = std::int32_t;
+
+  // Prepare input
+  const offset_t num_items = GENERATE_COPY(take(2, random(1, 1000000)), values({500, 1000000, 2000000}));
+  auto key_ranks           = make_shuffled_key_ranks_vector(num_items, CUB_SEED(2));
+  thrust::device_vector<key_t> keys_in_out(num_items);
+  thrust::transform(key_ranks.begin(), key_ranks.end(), keys_in_out.begin(), rank_to_key_op_t<offset_t, key_t>{});
+
+  // Perform sort
+  sort_keys(thrust::raw_pointer_cast(keys_in_out.data()), num_items, custom_less_op_t{});
+
+  // Verify results
+  thrust::host_vector<key_t> keys_expected(num_items);
+  auto key_ranks_it = thrust::make_counting_iterator(offset_t{});
+  thrust::transform(key_ranks_it, key_ranks_it + num_items, keys_expected.begin(), rank_to_key_op_t<offset_t, key_t>{});
+
+  REQUIRE(keys_expected == keys_in_out);
+}
+
+CUB_TEST("DeviceMergeSort::StableSortKeysCopy works and performs a stable sort when there are a lot sort-keys that "
+         "compare equal",
+         "[merge][sort][device]",
+         key_types)
+{
```

Comment:
question: from the merge sort perspective, tuples are similar to the custom type used for testing. Do you think we could get away with just testing the custom type here? This might simplify the code to just use gen, stable sort, compare.

Reply:
Agreed. I switched to `custom_type_t` instead of thrust::tuples. Made one test data generation to cause a lot of ties on sort keys and the other one with no special focus on ties.